course4/lecture4_en.md

```diff
@@ -0,0 +1,306 @@
+
+```moonbit
+fn f(info: PersonalInfo) -> (String, Int) { (info.name, info.age) }
+
+fn g(pair: (String, Int)) -> PersonalInfo { { name: pair.0, age: pair.1, }}
+```
+
+Feel free to verify this. Similarly, `PersonalInfo` is isomorphic to `(Int, String)`. You can try defining the corresponding mappings yourself.
+
+The key difference between tuples and structures lies in their compatibility. Tuples are *structural*, meaning they are compatible as long as the structure is the same – each field type corresponds one-to-one. For example, a function successfully accepts a tuple here. 
+
+```moonbit
+fn accept(tuple: (Int, String)) -> Bool {
+  true
+}
+let accepted: Bool = accept((1, "Yes"))
+```
+
+On the other hand, structures are *nominal*, meaning compatibility is based on the type name, and the internal order can be rearranged. In the first example, even though the structures are identical, the function cannot accept the structure because the type names are different. In the second example, the function can accept it because the types are the same even if the order of the fields is different.
+
+```moonbit
+struct A { val : Int ; other: Int }
+struct B { val : Int ; other: Int }
+fn accept(a: A) -> Bool {
+  true
+}
+let not_accepted: Bool = accept(({ val : 1, other : 2 }: B)) // DO NOT COMPILE
+let accepted: Bool = accept(({other: 2, val: 1}: A))
+```
+
+# Pattern Matching
+
+Pattern matching is another way to access tuples and structures.
+
+```moonbit
+fn head_opt(list: List[Int]) -> Option[Int] {
+  match list {
+    Nil => None
+    Cons(head, tail) => Some(head)
+  }
+}
+
+fn get_or_else(option_int: Option[Int], default: Int) -> Int {
+  match option_int {
+    None => default
+    Some(value) => value
+  }
+}
+```
+
+We have previously used pattern matching to inspect the structure of lists and Options. For instance, using `Nil` and `Cons` to match lists; `None` and `Some` to match Options. In fact, pattern matching can match values (logical values, numbers, characters, strings) as well as constructors. 
+
+
+```moonbit
+fn is_zero(i: Int) -> Bool {
+  match i {
+    0 => true
+    1 | 2 | 3 => false
+    _ => false
+  }
+}
+```
+
+In the examples above, we matched numbers. Here we use the pipe symbol (the 'or' pattern) to simultaneously match multiple possible values. The underscore (_) is the wildcard to match all remaining cases. We can nest patterns in constructors, or bind corresponding structures with identifiers.
+
+```moonbit
+fn contains_zero(l: List[Int]) -> Bool {
+  match l {
+    Nil => false
+    Cons(0, _) => true
+    Cons(_, tl) => contains_zero(tl)
+  }
+}
+```
+
+In this example, the first `Cons` uses the number 0 to match lists starting with 0, and the second `Cons` uses a wildcard and an identifier `tl` to match remaining lists while binding the sublist to the identifier `tl` for further processing. The value of the current list is discarded by the wildcard.
```

Comment:
I think it is a bit imprecise to say 'the value of the current list'. Maybe the head of the current list or the first element of the current list?
Also, `Cons` does not match 'remaining lists' -> it just matches the whole list

Reply:
Yes, `Cons` matches the whole list and I know that. My narrative caused ambiguity here. I meant the second `Cons` will match the rest of the lists left by the first `Cons` case. How about this revision: 

In this example, the first `Cons` match lists starting with 0 with the literal number 0. The second `Cons` uses a wildcard to match other lists, while binding the sublist to the identifier `tl` for further processing. The head of the current list is discarded by the wildcard.

Reply:
Yes, this is better. Also I think you can just write it out as 

> the branch `Cons(0, _)`
> the branch `Cons(_, tl)`